src/esp/bindings/SensorBindings.cpp

```diff
@@ -37,6 +38,94 @@ esp::scene::SceneNode* nodeGetter(T& self) {
+                                 py::int_(self.specification()->resolution[1]),
+                                 py::int_(self.specification()->channels)),
+                                "dtype"_a = torch.attr("uint32"),
+                                "device"_a = torch.attr("device")(
+                                    py::str("cuda"), py::int_(gpuDevice))));
+      }
+#endif
+    } else {
+      if (self.specification()->sensorType == SensorType::Semantic) {
+        auto pyBuffer = py::array(py::buffer_info(
+            nullptr,          /* Pointer to data (nullptr -> ask NumPy to
+                                 allocate!) */
+            sizeof(uint32_t), /* Size of one item */
+            py::format_descriptor<uint32_t>::value, /* Buffer format
+                                                     */
+            2,                                      /* How many dimensions? */
+            {self.specification()->resolution[0],
+             self.specification()->resolution[1]}, /* Number of elements for
+                                                     each dimension */
+            {sizeof(uint32_t), sizeof(uint32_t)}
+            /* Strides for each dimension */
+            ));
+        py::setattr(handle, "__buffer", pyBuffer);
+      } else if (self.specification()->sensorType == SensorType::Depth) {
+        auto pyBuffer = py::array(py::buffer_info(
+            nullptr,       /* Pointer to data (nullptr -> ask NumPy to
+                              allocate!) */
+            sizeof(float), /* Size of one item */
+            py::format_descriptor<float>::value, /* Buffer format */
+            2,                                   /* How many dimensions? */
+            {self.specification()->resolution[0],
+             self.specification()->resolution[1]}, /* Number of elements for
+                                                     each dimension */
+            {sizeof(float), sizeof(float)} /* Strides for each dimension */
+            ));
+        py::setattr(handle, "__buffer", pyBuffer);
+      } else {
+        auto pyBuffer = py::array(py::buffer_info(
+            nullptr,         /* Pointer to data (nullptr -> ask NumPy to
+                                allocate!) */
+            sizeof(uint8_t), /* Size of one item */
+            py::format_descriptor<uint8_t>::value, /* Buffer format */
+            3,                                     /* How many dimensions? */
+            {self.specification()->resolution[0],
+             self.specification()->resolution[1],
+             self.specification()->channels}, /* Number of elements
+                                                 for each dimension */
+            {sizeof(uint8_t), sizeof(uint8_t), sizeof(uint8_t)}
+            /* Strides for each dimension */
+            ));
+        py::setattr(handle, "__buffer", pyBuffer);
+      }
+    }
+  }
+  return py::getattr(handle, "__buffer");
+};
```

Comment:
Wait, is this all new code? If it is, how was this handled previously? I'm confused.

In my opinion Habitat isn't a place to do such low level buffer protocol operations, there needs to be *extensive* testing for such pieces of code because a lot of things can go wrong -- and that's why it's all already implemented in magnum bindings. Zero-copy array transfer via a buffer protocol was originally done through magnum's image views, did that change now? Why does the old way not work anymore?

Reply:
It's an attempt to deduplicate and generalize this Python class into one
that uses pure C++ calls:
https://github.com/facebookresearch/habitat-sim/blob/3c7d566c00d6c0c15165cbdf9c4fc791c44397de/habitat_sim/simulator.py#L476

Reply:
To follow up on this, given that we still need something on the python side for noise models, this can also be done in python.